graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,172 @@
 # Copyright 2019-present Kensho Technologies, LLC.
-from collections import namedtuple
+from dataclasses import dataclass, field
+from typing import List, NamedTuple, Tuple
+
+from graphql import DocumentNode
 
 from ..ast_manipulation import get_only_query_definition, get_only_selection_from_ast
-from ..cost_estimation.helpers import is_int_field_type, is_uuid4_type
+from ..cost_estimation.helpers import is_uuid4_type
+from ..cost_estimation.int_value_conversion import field_supports_range_reasoning
 from ..exceptions import GraphQLError
+from ..schema.schema_info import QueryPlanningSchemaInfo
+
+
+@dataclass
+class PaginationAdvisory(object):
+    message: str
+
+
+@dataclass
+class PaginationFieldNotSpecified(PaginationAdvisory):
+    vertex_name: str
+    message: str = field(init=False)
+
+    def __post_init__(self):
+        """Initialize a human-readable message."""
+        self.message = (
+            "Specifying a pagination field for vertex {} in the QueryPlanningSchemaInfo "
+            "would have made this vertex eligible for pagination, and enabled a better "
+            "pagination plan.".format(self.vertex_name)
+        )
+
+
+@dataclass
+class InsufficientQuantiles(PaginationAdvisory):
+    vertex_name: str
+    field_name: str
+    current_resolution: int
+    desired_resolution: int
+    message: str = field(init=False)
+
+    def __post_init__(self):
+        """Initialize a human-readable message."""
+        self.message = (
+            "Pagination would have been more successful if more quantiles were provided "
+            "for {}.{}. Currently there are {}, ideally there should be {}".format(
+                self.vertex_name, self.field_name, self.current_resolution, self.desired_resolution
+            )
+        )
+
 
+class VertexPartition(NamedTuple):
```

Comment:
`VertexPartitionPlan`, perhaps, to make it clear it is not the actual partition?